### PR TITLE
Force OpenSSH to use SSH_ASKPASS if a password is requested

### PR DESCRIPTION
**Pull request by SlugFiller**
## Proposed changes

Minimalist alternative to #11182. This adds an environment variable that forces SSH_ASKPASS to be used even if a terminal is available. Requires OpenSSH 8.4 or higher. Has no effect on older versions of OpenSSH.

(The version that comes with Windows 10 is, unfortunately, 8.1. However, 9.2 is available via winget)

### Before

Inputting a password is done via the console emulator.

### After

Inputting a password is done via the `GitExtSshAskPass.exe` popup.

## Test methodology 

- Performed fetch from a remote server
- Inputted the password in the popup

## Test environment(s) 

- GIT 2.41.0.windows.1
- Windows 10.0.19045

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitCommands/Git/EnvironmentConfiguration.cs [hunk 1]
```diff
@@ -60,6 +66,11 @@ public static void SetEnvironmentVariables()
                 Env.SetEnvironmentVariable("SSH_ASKPASS", "ssh-askpass");
             }
 
+            if (!string.IsNullOrEmpty(Env.GetEnvironmentVariable("SSH_ASKPASS")))
```
Comment (mstv) [inline GitCommands/Git/EnvironmentConfiguration.cs:69]: ❓ If intentionally, add a comment please. Otherwise
```suggestion
            if (string.IsNullOrEmpty(Env.GetEnvironmentVariable("SSH_ASKPASS")))
```

Comment (SlugFiller) [reply GitCommands/Git/EnvironmentConfiguration.cs:69]: No, it adds `SSH_ASKPASS_REQUIRE` only if `SSH_ASKPASS` is present, since it would be meaningless if it is absent (No program to call). Technically, the only case where this `if` isn't entered is if `File.Exists(sshAskPass)` is false above.

Comment (mstv) [reply GitCommands/Git/EnvironmentConfiguration.cs:69]: I just did not notice the suffix "_REQUIRE".

**Review by mstv — COMMENTED** [review]
+1 Looks sensible.
Though I am not familiar with the details. Is there a risk that these environment variables break it for other users?

---

Anyway, I want to use #11182 for the process dialog without ConEmu in order to get rid of loading times and of additional console windows.

**Comment by SlugFiller** [comment]
On older versions of OpenSSH, this would be ignored. On PuTTY, this would be ignored. And while I would prefer that the environment variable would only be set on the child process, I did verify that it doesn't leak into the parent process even if using `dotnet run`, so it shouldn't influence program use outside of GitExtensions.

I would also prefer #11182 over this, if only for the lack of a need to install a specific version of OpenSSH, but it's been in limbo, and I've received no comments, questions, or suggestions for changes on it in over 10 days. Which might not sound like a lot, but this is a blocker for me. I'm holding back migrations from Mercurial to Git over this.

Incidentally, the two are mutually exclusive, as popping up SSH_ASKPASS removes the ability to type in the password directly via the terminal.

**Review by RussKie — APPROVED** [review]
👍

**Comment by RussKie** [comment]
> Anyway, I want to use #11182 for the process dialog without ConEmu in order to get rid of loading times and of additional console windows.

Are you not using the ConEmu?

**Comment by RussKie** [comment]
Let's take this one as it looks simpler, and it'll unblock you in the interim. When we get more traction on https://github.com/gitextensions/gitextensions/pull/11182 we can undo this, if required.

**Comment by SlugFiller** [comment]
[This comment](https://stackoverflow.com/questions/77494275/in-git-extensions-for-windows-4-2-0-how-can-we-disable-the-command-line-wrapper#comment136622968_77494275) actually gives the answer perfectly. Since this is inside bash, it's entirely possible to set an environment variable. The down side is that doing it this way only applies to the one session.

I admit that I did not expect the temporary stripped environment made for git commands to apply inside the git bash pop-up, which uses a completely different type of wrapper. Some form of separating the two cases might be required.

I don't see how special characters should be an issue, though. The GUI textbox should handle them just as gracefully, if not more, than a command line terminal. This, in and of itself, deserves testing and correction. Assuming the issue is in the ask-pass client, and not in Git itself.

Finally, it might be worthwhile to make askpass itself configurable. Currently, it forces the use of `GitExtSshAskPass.exe` on Windows, even if an environment variable exists. The only way to disable it is to delete the file `GitExtSshAskPass.exe` itself. Given that without `SSH_ASKPASS_REQUIRE`, askpass doesn't work anywhere anyway, this is sufficient for restoring the previous behavior, and might work as an accepted answer for that particular StackOverflow question. But not really a resolution for the underlying issue.

TL;DR This is revealing multiple issues, with multiple approaches to resolution. All angles need to be considered.

**Comment by RussKie** [comment]
This change seems to be causing quite a bit of distress to users, and I'm starting to think we need to roll it back and figure out the way to make it less intrusive.

**Comment by SlugFiller** [comment]
I've only seen one user complain, and the issue there is with the wrapper, not this PR. The wrapper has thus far been untested, because even though code existed to run it, that code was not working on Windows, due to quirks with OpenSSH and the console emulator implementation. And also the fact that, apparently, none of the maintainers log into any SSH server using a password.

I explained it above, but the path forward is to do one or more of the following:
1. Ensure the wrapper works with non-ASCII characters. Possibly: Allow the user to configure the chosen encoding.
2. Ensure the wrapper is not set in the Git bash/cmd/powershell launched from the tool bar. This one is questionable. UX-wise, it makes sense for an interactive shell to not pop-up a window, and instead ask for a password inside the terminal. While it makes sense for a non-interactive shell to ask for a password in a separate GUI. But in terms of inputting non-ASCII passwords, the problem remains when doing push/pull. Also this requires refactoring the environment abstraction to not [directly make modifications via `System.Environment`](https://github.com/gitextensions/gitextensions/blob/44be238d28e5c68ee11bca3f67351d7383777e65/GitCommands/EnvironmentAbstraction.cs#L59) and instead only provide an environment to [spawned processes that need it](https://github.com/gitextensions/gitextensions/blob/44be238d28e5c68ee11bca3f67351d7383777e65/GitCommands/Git/Executable.cs#L39)
3. Create a new setting to complete disable `SSH_ASKPASS`, and/or respect the user's environment settings, on Windows. On Linux, the user's environment is already respected. But on Windows, so long as the wrapper exists, it is used. See [here](https://github.com/gitextensions/gitextensions/blob/44be238d28e5c68ee11bca3f67351d7383777e65/GitCommands/Git/EnvironmentConfiguration.cs#L55).
4. Same as 3, but only for `SSH_ASKPASS_REQUIRE`. This is more than a bit ugly, since the fact that `SSH_ASKPASS` is usually ignored, is pretty much a bug, not a feature. Setting it only to have it ignored shouldn't happen.

1 is a must, but the others require debate on whether to do them or not.
